File: backend/scripts/check_debug_code.py
#!/usr/bin/env python3
"""
Check for debug code in Python files
Prevents print statements and debugger calls from being committed to production code
"""

import re
import subprocess
import sys
from pathlib import Path

DEBUG_PATTERNS = [
    (r"\bprint\s*\(", "print() statement"),
    (r"\bbreakpoint\s*\(", "breakpoint() call"),
    (r"\bpdb\.set_trace\s*\(", "pdb.set_trace() call"),
    (r"\bipdb\.set_trace\s*\(", "ipdb.set_trace() call"),
    (r"import\s+pdb", "pdb import"),
    (r"import\s+ipdb", "ipdb import"),
]

ALLOWED_PATHS = [
    "tests/",
    "test_",
    "scripts/",
    "migrations/",
    "alembic/",
    "scaffolding/",
    "auth_status_report.py",
    "check_dependencies.py",
]


def is_allowed_file(filepath):
    """Check if file is allowed to have debug code"""
    return any(pattern in str(filepath) for pattern in ALLOWED_PATHS)


def check_file(filepath, content):
    """Check a single file for debug patterns"""
    issues = []

    if is_allowed_file(filepath):
        return issues

    lines = content.split("\n")

    for pattern, name in DEBUG_PATTERNS:
        for line_num, line in enumerate(lines, 1):
            # Skip comments
            if line.strip().startswith("#"):
                continue

            if re.search(pattern, line):
                issues.append(
                    {
                        "file": filepath,
                        "line": line_num,
                        "pattern": name,
                        "code": line.strip(),
                    }
                )

    return issues


def main():
    try:
        # Get staged Python files
        result = subprocess.run(
            ["git", "diff", "--cached", "--name-only", "--diff-filter=ACM"],
            capture_output=True,
            text=True,
            check=True,
        )

        staged_files = [
            f
            for f in result.stdout.split("\n")
            if f.endswith(".py") and f.startswith("backend/")
        ]

        if not staged_files:
            sys.exit(0)

        all_issues = []

        for file in staged_files:
            try:
                # Get staged content
                result = subprocess.run(
                    ["git", "show", f":{file}"],
                    capture_output=True,
                    text=True,
                    check=True,
                )

                issues = check_file(file, result.stdout)
                all_issues.extend(issues)

            except subprocess.CalledProcessError:
                # File might be deleted or binary, skip
                continue

        if all_issues:
            print("\n❌ Debug code detected in staged Python files:\n")

            for issue in all_issues:
                print(f"  {issue['file']}:{issue['line']}")
                print(f"    {issue['pattern']}: {issue['code']}")
                print("")

            print("🚫 Please remove debug code before committing.\n")
            print("💡 Tip: Use Python logging module in production code.\n")

            sys.exit(1)

        sys.exit(0)

    except Exception as e:
        print(f"Error checking debug code: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: frontend/scripts/check-debug-code.js
#!/usr/bin/env node
/**
 * Check for debug code in staged files
 * Prevents console.log, debugger, and other debug statements from being committed
 */

import { execSync } from "child_process";

const DEBUG_PATTERNS = [
  {
    pattern: /console\.(log|debug|info|warn)\(/g,
    name: "console.log/debug/info/warn",
  },
  { pattern: /debugger;/g, name: "debugger statement" },
  { pattern: /\.only\(/g, name: "test.only or describe.only" },
  { pattern: /\.skip\(/g, name: "test.skip or describe.skip" },
];

const ALLOWED_FILES = [
  /\.test\.(ts|tsx|js|jsx)$/, // Test files
  /\.spec\.(ts|tsx|js|jsx)$/, // Spec files
  /\.stories\.(ts|tsx|js|jsx)$/, // Storybook files
  /scripts\//, // Script files
  /\.config\.(ts|js)$/, // Config files
];

function checkFile(filePath, content) {
  const issues = [];

  // Skip allowed files
  if (ALLOWED_FILES.some((pattern) => pattern.test(filePath))) {
    return issues;
  }

  // Check for debug patterns
  DEBUG_PATTERNS.forEach(({ pattern, name }) => {
    const matches = content.match(pattern);
    if (matches) {
      const lines = content.split("\n");
      lines.forEach((line, index) => {
        if (pattern.test(line)) {
          issues.push({
            file: filePath,
            line: index + 1,
            pattern: name,
            code: line.trim(),
          });
        }
      });
    }
  });

  return issues;
}

function main() {
  try {
    // Get staged files
    const stagedFiles = execSync(
      "git diff --cached --name-only --diff-filter=ACM",
      {
        encoding: "utf8",
      },
    )
      .split("\n")
      .filter((file) => file.match(/\.(ts|tsx|js|jsx)$/));

    if (stagedFiles.length === 0) {
      process.exit(0);
    }

    let allIssues = [];

    stagedFiles.forEach((file) => {
      try {
        const content = execSync(`git show :${file}`, { encoding: "utf8" });
        const issues = checkFile(file, content);
        allIssues = allIssues.concat(issues);
      } catch {
        // File might be deleted or binary, skip
      }
    });

    if (allIssues.length > 0) {
      console.error("\n❌ Debug code detected in staged files:\n");

      allIssues.forEach((issue) => {
        console.error(`  ${issue.file}:${issue.line}`);
        console.error(`    ${issue.pattern}: ${issue.code}`);
        console.error("");
      });

      console.error("🚫 Please remove debug code before committing.\n");
      console.error(
        "💡 Tip: Use a proper logging library in production code.\n",
      );

      process.exit(1);
    }

    process.exit(0);
  } catch (error) {
    console.error("Error checking debug code:", error.message);
    process.exit(1);
  }
}

main();
